Library/Homebrew/cmd/update-bash.sh

```diff
@@ -2,22 +2,38 @@ brew() {
   "$HOMEBREW_BREW_FILE" "$@"
 }
 
+git() {
+  "$HOMEBREW_GIT" "$@"
+}
+
 which_git() {
-  local which_git
   local active_developer_dir
 
-  which_git="$(which git 2>/dev/null)"
-  if [[ -n "$which_git" && "/usr/bin/git" = "$which_git" ]]
+  if [[ -n "$HOMEBREW_GIT" && -x "$HOMEBREW_GIT" ]]
+  then
+    echo "$HOMEBREW_GIT"
+    return
+  fi
+
+  if [[ -n "$GIT" && -x "$GIT" ]]
+  then
+    HOMEBREW_GIT="$GIT"
+    echo "$HOMEBREW_GIT"
+    return
+  fi
+
+  HOMEBREW_GIT="$(which git 2>/dev/null)"
```

Comment:
I think the control flow in this part should be more along the lines of

```shell
if [[ -n "$HOMEBREW_GIT" ]]
then
  : # We'll take the user-provided `HOMEBREW_GIT` directly.
elif [[ -n "$GIT" ]]
then
  HOMEBREW_GIT="$GIT"
else
  HOMEBREW_GIT="$(which git 2>/dev/null)"
fi
```

or condensed to a single line (but maybe deemed too obscure)

```shell
: "${HOMEBREW_GIT:=${GIT:-$(which git 2>/dev/null)}}"
```

followed by the special handling of `/usr/bin/git` and the active developer directory (we should be doing this no matter where the `git` path came from), finally followed by a check if the determined path exists and is executable.